lib/src/merged_tree.rs

```diff
@@ -971,6 +974,235 @@ impl Iterator for TreeDiffIterator<'_> {
+        &mut self,
+        dir: RepoPath,
+        tree_diff: BackendResult<(MergedTree, MergedTree)>,
+    ) {
+        let (tree1, tree2) = match tree_diff {
+            Ok(trees) => trees,
+            Err(err) => {
+                self.items.insert(DiffStreamKey::normal(dir), Err(err));
+                return;
+            }
+        };
+
+        for basename in merged_tree_basenames(&tree1, &tree2) {
+            let value_before = tree1.value(basename);
+            let value_after = tree2.value(basename);
+            if value_after != value_before {
+                let path = dir.join(basename);
+                let before = value_before.to_merge();
+                let after = value_after.to_merge();
+                if (before.is_tree() || after.is_tree()) && !self.matcher.visit(&path).is_nothing()
+                {
+                    let before_tree_future = Self::tree(
+                        self.store.clone(),
+                        self.legacy_format_before,
+                        path.clone(),
+                        before.clone(),
+                    );
+                    let after_tree_future = Self::tree(
+                        self.store.clone(),
+                        self.legacy_format_after,
+                        path.clone(),
+                        after.clone(),
+                    );
+                    let both_trees_futures =
+                        async { futures::try_join!(before_tree_future, after_tree_future) };
+                    self.pending_trees
+                        .push_back((path.clone(), Box::pin(both_trees_futures)));
+                }
+                if self.matcher.matches(&path) {
+                    self.items
+                        .insert(DiffStreamKey::normal(path), Ok((before, after)));
+                }
+            }
+        }
+    }
+}
+
+impl Stream for TreeDiffStreamImpl<'_> {
+    type Item = (RepoPath, BackendResult<(MergedTreeValue, MergedTreeValue)>);
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut need_items = false;
+        while !(self.items.is_empty() && self.pending_trees.is_empty()) {
+            // Go through all pending tree futures and poll them.
+            let mut pending_index = 0;
+            while pending_index < self.pending_trees.len()
+                && (need_items
+                    || pending_index < self.max_concurrent_reads
+                    || self.items.len() < self.max_queued_items)
+            {
+                let (_, future) = &mut self.pending_trees[pending_index];
+                if let Poll::Ready(tree_diff) = future.as_mut().poll(cx) {
+                    let (dir, _) = self.pending_trees.remove(pending_index).unwrap();
+                    let key = DiffStreamKey::normal(dir);
+                    let Some(Ok((before, after))) = self.items.remove(&key) else {
+                        unreachable!();
+                    };
```

Comment:
nit: could be `let (before, after) = ...unwrap().unwrap()`, and could have some description why unwrapping is okay.

Actually, it panicked with path "lib/src" since "lib" is visited but doesn't match.

Reply:
Good catch! We didn't have any testing for diffs with matcher so I added a commit for that. An entry in `self.items` should have been added when we added one in `self.pending_trees`. I've fixed that now. I've also updated `add_dir_diff_items()` quite a bit to better respect the matcher.